salt/states/file.py

```diff
@@ -557,7 +558,26 @@ def process_symlinks(filenames, symlinks):
             managed_directories.add(mdest)
             keep.add(mdest)
 
-    return managed_files, managed_directories, managed_symlinks, keep
+    # Sets are randomly ordered. We need to use a list so we can make sure
+    # symlinks are always at the end. This is necessary because the file must
+    # exist before we can create a symlink to it. See issue:
+    # https://github.com/saltstack/salt/issues/64630
+    new_managed_files = list(managed_files)
+    # Now let's move all the symlinks to the end
+    for link_src_relpath, _ in managed_symlinks:
+        for file_dest, file_src in managed_files:
+            # We need to convert relpath to fullpath. We're using pathlib to
+            # be platform-agnostic
+            symlink_full_path = pathlib.Path(f"{name}{os.sep}{link_src_relpath}")
+            file_dest_full_path = pathlib.Path(file_dest)
+            if symlink_full_path == file_dest_full_path:
```

Comment:
Shouldn't we be checking hte symlink target instead of the symlink source?
For example the link is something like

Link:
```
rel_path/mylink -> /path/to/target
```
so  
`link_src_relpath == "rel_path/mylink"`  
`file_dest = "rel_path/to/target"`  
`ltarget == /path/to/target`

meaning we need to check that `file_dest_full_path == /path/to/target`. I assume the link target can be relative or absolute so we would need to resolve it first with something like `Path.resolve()`

Reply:
Never mind; I confused myself. I forgot we want those target files to be at before the symlink. This is trying to get the determine the files that are symlinks since the list contains both the non-symlink files and the symlinks and then move those to the end.

Reply:
This is all very complicated. Ideally we could just have a separate `cp` function or flag to `cp.list_master` that returned just the non-symlink files.

Reply:
I thought about this a little more driving home from work today and realized that the code is already iterating over all the symlinks and removing the files that starts with the symlink. We can do this processing to move the files to the end in the same loop and avoid an extra iteration over all the symlinks. Or we can just go ahead and remove the symlink from the files list there (although it would still be ideal to have a server side function that only returns the files without symlinks) and then update the code later on to expect the symlinks to only be in the symlinks collection and not in the files collection. I haven't looked at what the code is currently doing with the symlinks in the files list but I would think it would not be much; having the symlinks and files separate seems like it would make the code easier to think about.